### PR TITLE
Docker 1.12

### PR DESCRIPTION
**Pull request by gianarb**

### CODE REVIEW

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+tags: [docker]
+summary: "Docker 1.12 adds different new features around orchestration, scaling
+and deployment, in this article I am happy to share sometest did this this new
+version"
+priority: 0.6
+changefreq: yearly
+---
+
+During the DockerCon 2016 docker announced Docker 1.12 release.
+One of the news around the new version is the orchestration system builded directly
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `the big news stories around`... `built directly`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+inside the engine, this feature allow us to use swarm
+without catch it from outside, it’s now a feature provided by Docker directly.
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: not sure `catch it` is best here... maybe `without installing it separately`?

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+It means stop to fetch a new project but all that you need is already provided by docker.
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: Maybe drop this sentence? Its repeating what has been said already? :)

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown (hunk 2 above)
Comment (gianarb) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: yes, it sounds less rude! :)

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+Now we have a new set of commands that allow us to orchestrate containers
+across a cluster.
+
+{% highlight bash %}
+docker swarm
+docker node
+docker service
+{% endhighlight %}
+
+All this commands are focused to increase our ability to orchestrate our
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `All these` ... `focused on increasing our`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+echo "sudo /etc/init.d/docker stop && \
+    curl https://test.docker.com/builds/Linux/x86_64/docker-1.12.0-rc2.tgz | \
+    tar xzf - && sudo mv docker/* /usr/local/bin && \
+    rm -rf docker/ && sudo /etc/init.d/docker start" | \
+    docker-machine ssh sw02 sh -
+docker-machine ssh sw02 docker swarm join $(docker-machine ip sw01):2377
+{% endhighlight %}
+
+ another Captain wrote this script that I just updated to work
+with the public Docker 1.12-rc2 we can use this script to create a cluster with
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `...Docker 1.12-rc2. We...`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+virtual box ready to be used.  After this script you can see the number of
+workers and masters, in this case your one and one.
+
+{% highlight bash %}
+$ docker node ls
+{% endhighlight %}
+
+Docker 1.12 has a build in set of primitive functions to orchestrate your containers just
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `built-in`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+like a summary the main commands that you must run to create a cluster are
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `a summary. The ...`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+docker-machine are
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: remove this line?

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+
+{% highlight bash %}
+## On the master to start your cluster
+$ docker swarm init --listen-addr <master-IP(this ip)>:2377
+## on each note to add it into the cluster
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `node`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+$ docker swarm <master-ip>:2377
+{% endhighlight %}
+
+<img class="img-responsive" alt="Docker Swarm architecture" src="/img/posts/swarm_arch.png">
+
+If you are not confident with docker swarm this is the architecture, this graph
+is provided by Docker Inc. explains really well the design around this project,
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `and explains` ... `around this project. The principal...`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown (hunk 8 above)
Comment (gianarb) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: 👍

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+the principal actors are managers and workers, managers are the brains of the
+system they dispatches schedules and remembers services and containers, workers
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `system, they dispatch schedules and remember services and containers. Workers`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+execute this commands.
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: these

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+
+<img class="img-responsive" alt="Docker Swarm architecture" src="/img/posts/swarm_arch.png">
+
+If you are not confident with docker swarm this is the architecture, this graph
+is provided by Docker Inc. explains really well the design around this project,
+the principal actors are managers and workers, managers are the brains of the
+system they dispatches schedules and remembers services and containers, workers
+execute this commands. 
+
+All cluster is secure because each node has a proper TLS
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `clusters are secure`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+identity and all communication are encrypted end to endby default with a
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `is encrypted end to end by default with automatic key...`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+automatic key rotation in order to increase the secury around the keys use in
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `the security around the keys used in`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+the cluster.  
+
+[Raft](https://raft.github.io/) is the consensual protocol used to distribute
+message around the cluster and check the number of nodes, it’s complex
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `messages` .. `number of nodes. It's a complex...`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+algorithm but really interested I have in plan another article about it but the
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `really interesting. I have...`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+offical site contains a lot of details about it.
+
+We already saw the concept of service in docker-compose it’s a single or a
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `services`?

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+group of containers to describe your ecosystem, you can scale a specific
+service or orchestrate it across your cluster. It’s the same here, you have not
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `you don't have`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 20]
```diff
@@ -0,0 +1,219 @@
...
+a specification file like compose at the moment but you can anyway run a bunch
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `but anyway you can run...`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+of commands to create your service.
+
+{% highlight bash %}
+$ docker service create --name helloworld --replicas 1 alpine ping docker.com
+{% endhighlight %}
+
+With this example we push up a new service helloworld it has one container from
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `helloworld. It has...`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 22]
```diff
@@ -0,0 +1,219 @@
...
+docker service ls
+{% endhighlight %}
+
+To watch all our services, we can also inspect a service
+
+{% highlight bash %}
+docker service inspect <container_id>
+{% endhighlight %}
+
+There is a new concept, when you run a service you are also creating a task
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `creating a task,`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 23]
```diff
@@ -0,0 +1,219 @@
...
+this task represent the container/s under your service in this case we have
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `represents`... `under your service, in this case...`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 24]
```diff
@@ -0,0 +1,219 @@
...
+docker service tasks helloworld
+{% endhighlight %}
+
+You can scale your service
+
+{% highlight bash %}
+docker service scale helloworld=10
+{% endhighlight %}
+
+Now you can see 10 tasks that are running and if you inspect one of the there
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `one of them`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 25]
```diff
@@ -0,0 +1,219 @@
...
+is the ID of your container and you can for example follow logs
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `can, for example, `

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 26]
```diff
@@ -0,0 +1,219 @@
...
+                "PID": 3627
+            }
+        },
+        "DesiredState": "running"
+    }
+]
+22:17 $ docker logs -f 6fhfse4it8lwzlsk1t5sd5jbk
+{% endhighlight %}
+
+At this point It is a normal container and it’s running on your cluster.
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `it is`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 27]
```diff
@@ -0,0 +1,219 @@
...
+Well I tried to explain the main concept around this big feature provided by Docker 1.12, the last example is just to cover the DNS topic, I created an application that serve an http server and print the current IP. Each server has an internal load balancer that dispatch traffic in round robin between the different tasks, in this way it’s totally transparent, you can just resolve your service with a normal URL, docker do the rest for you.
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `DNS topic. I created...`

Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `DNS topic. I created...`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 28]
```diff
@@ -0,0 +1,219 @@
...
+
+{% highlight bash %}
+$. docker service create —-name micro —-replicas 10 —-publish 8000/tcp gianarb/micro
+{% endhighlight %}
+
+[Micro](https://github.com/gianarb/micro) is an application that exposes an http server on port 8000 and print the current ip, now we have 10 tasks with this service, to grab the current entry point for our service we can inspect it and search this information:
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `this service. To grab...`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 29]
```diff
@@ -0,0 +1,219 @@
...
+                {
+                    "NetworkID": "890fivvc6od3pa4rxd281lobb",
+                    "Addr": "10.255.0.5/16"
+                }
+            ]
+       }
+...
+{% endhighlight %}
+
+In this case our publish port is 3000 we can call <ip>:3000 to resolve our
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `In this case, our published port is 3000, we can...`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 30]
```diff
@@ -0,0 +1,219 @@
...
+service, if you try to do multi request you can see your IP chance because the
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `changes`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 31]
```diff
@@ -0,0 +1,219 @@
...
+internal DNS is calling different containers.
+
+This is just an overview about the features but there are other powerfull news
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `powerful`

File: _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown [hunk 32]
```diff
@@ -0,0 +1,217 @@
+---
+layout: post
+title:  "Docker 1.12 orchestration built-in"
+date:   2016-06-20 10:08:27
+categories: [post]
+img: /img/docker.png
+tags: [docker]
+summary: "Docker 1.12 adds different new features around orchestration, scaling
+and deployment, in this article I am happy to share sometest did this this new
```
Comment (garethpelly) [inline _posts/2016-06-20-docker-1-12-orchestration-build-in.markdown]: `some tests I did with this new`

**Pull request merged**